### PR TITLE
Deprecate functions `neighbor`, `runningAccumulate`, `runningDifferenceStartingWithFirstValue`, `runningDifference`

### LINKED ISSUE #51901: Problem with query plan optimizer yielding incorrect results
After upgrading ClickHouse server, one of our queries started returning incorrect results due to a problem with query plan optimizations. The issue was first introduced in `v22.4` and I believe it's related to this PR: #35623.

Consider the following query:

```SQL
SELECT
    Dim,
    CumulativeSum
FROM
(
    SELECT
        Dim,
        runningAccumulate(SumState) AS CumulativeSum
    FROM
    (
        SELECT
            Dim,
            sumState(DimCount) AS SumState
        FROM
        (
            SELECT
                c1 AS Dim,
                count() as DimCount
            FROM VALUES(9, 5, 5, 2, 1, 1)
            GROUP BY Dim
        )
        GROUP BY Dim
        ORDER BY Dim DESC
    )
)
ORDER BY Dim ASC
```

The correct results of this query should be:

```
┌─Dim─┬─CumulativeSum─┐
│   1 │             6 │
│   2 │             4 │
│   5 │             3 │
│   9 │             1 │
└─────┴───────────────┘
```

But starting from v22.4, ClickHouse query plan optimizer decides to calculate `runningAccumulate` after sorting which ruins the cumulative sum and gives incorrect results.

[ClickHouse Playground v22.3](https://fiddle.clickhouse.com/c77e6c20-830b-413b-ad78-3d9e59707a4f) (gives correct results)
[ClickHouse Playground v22.4](https://fiddle.clickhouse.com/c2a08315-2850-425e-bf9b-189ff3200f0b) (gives incorrect results)

Note that by setting `query_plan_enable_optimizations=0`, the query gives correct results which indicates the issue is related to the query plan optimizer.

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Usage of functions `neighbor`, `runningAccumulate`, `runningDifferenceStartingWithFirstValue`, `runningDifference` deprecated (because it is error-prone). Proper window functions should be used instead. To enable them back, set `allow_deprecated_functions=1`.

---

Closes: #51901, #16484, #59416, #60441, #57287

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 64ee0090cbc447e50d725e5f6f895d369c1e0996 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63132/64ee0090cbc447e50d725e5f6f895d369c1e0996/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by alexey-milovidov — APPROVED** [review]

**Comment by den-crane** [comment]
AFAIK 
by this PR you can close #51901, #16484, #59416, #60441, #57287

**Comment by den-crane** [comment]
It closed only the first issue. It does not understand multiple comma separated issues.

Not OK close: #51901, #16484, #59416, #60441, #57287
OK: close: #51901, close: #16484, close: #59416, close: #60441, close: #57287

**Comment by Slach** [comment]
which  functions i should use instead of  `neighbor` and `runningDifference`?

what exactly means "proper window function" in docs?

**Comment by Algunenano** [comment]
> what exactly means "proper window function" in docs?

It means https://clickhouse.com/docs/en/sql-reference/window-functions. The deprecated functions were trying to act as window functions without being window functions, which meant lots of reports because they didn't work as people expected them to work due to unstable execution order.

**Comment by den-crane** [comment]
@Slach https://kb.altinity.com/altinity-kb-queries-and-syntax/lag-lead/#using-laginframeleadinframe-starting-from-clickhouse-214

File: src/Core/Settings.h [hunk 1]
```diff
@@ -912,6 +912,7 @@ class IColumn;
     M(Int64, ignore_cold_parts_seconds, 0, "Only available in ClickHouse Cloud. Exclude new data parts from SELECT queries until they're either pre-warmed (see cache_populated_by_fetch) or this many seconds old. Only for Replicated-/SharedMergeTree.", 0) \
     M(Int64, prefer_warmed_unmerged_parts_seconds, 0, "Only available in ClickHouse Cloud. If a merged part is less than this many seconds old and is not pre-warmed (see cache_populated_by_fetch), but all its source parts are available and pre-warmed, SELECT queries will read from those parts instead. Only for ReplicatedMergeTree. Note that this only checks whether CacheWarmer processed the part; if the part was fetched into cache by something else, it'll still be considered cold until CacheWarmer gets to it; if it was warmed, then evicted from cache, it'll still be considered warm.", 0) \
     M(Bool, iceberg_engine_ignore_schema_evolution, false, "Ignore schema evolution in Iceberg table engine and read all data using latest schema saved on table creation. Note that it can lead to incorrect result", 0) \
+    M(Bool, allow_deprecated_functions, false, "Allow usage of deprecated functions", 0) \
```
Comment (Algunenano) [inline src/Core/Settings.h:915]: This name is too generic as might cause issues in the future when we deprecate other functions and people thing of reusing the setting, as then we won't have a way to maintain compatibility. WDTY of changing it to something like `allow_deprecated_window_functions` instead?

Comment (Algunenano) [reply src/Core/Settings.h:915]: Or even call them `unproper` / `bad` _window_functions maybe?

Comment (nickitat) [reply src/Core/Settings.h:915]: I don't mind but atst not sure that we need so granular settings

**Pull request merged**